### ISSUE TITLE
Marathon should set a non-default filter when declining offers

### ISSUE THREAD

**Issue by ConnorDoyle** [mesosphere/marathon#1931]
**Use case:** _As a cluster operator I would like to run many frameworks (possibly multiple Marathon instances) on the same Mesos cluster.  I'd like to mitigate offer starvation; that is, schedulers with work to do should receive offers from the Mesos master._

From the SchedulerDriver documentation, Marathon can use both the overloaded definition of `declineOffer` to set filters:

```java
  /**
   * Declines an offer in its entirety and applies the specified
   * filters on the resources (see mesos.proto for a description of
   * Filters). Note that this can be done at any time, it is not
   * necessary to do this within the {@link Scheduler#resourceOffers}
   * callback.
   *
   * @param offerId The ID of the offer to be declined.
   * @param filters The filters to set for any remaining resources.
   *
   * @return        The state of the driver after the call.
   *
   * @see OfferID
   * @see Filters
   * @see Status
   */
  Status declineOffer(OfferID offerId, Filters filters);
```

and clear them later on.

```
  /**
   * Removes all filters, previously set by the framework (via {@link
   * #launchTasks}). This enables the framework to receive offers
   * from those filtered slaves.
   *
   * @return    The state of the driver after the call.
   *
   * @see Status
   */
  Status reviveOffers();
```

**Proposal:** _When the task queue becomes empty, decline subsequent offers with a long timeout (say, max double value).  When the task queue becomes non-empty, invoke reviveOffers. (per offline conversation with @kolloch, @aquamatthias, and @gkleiman)._

**Comment by adam-mesos** [comment]
Couldn't you also just abort() the SchedulerDriver and then create a new one (sends ReregisterFrameworkMessage) when you have tasks again? This abort() results in a DeactivateFrameworkMessage, which will temporarily deactivate the framework in the Mesos master allocator, preventing any new offers, while still leaving tasks running. Since abort() doesn't actually terminate the SchedulerDriver, you might still get status updates, etc. (I'd have to double-check). Reregistering will start the offers again.
Also note that other users have found that an excess of filters (one per each of 100 frameworks per each of 1000 nodes) can slow down the master's allocation/offer cycle. So I'd be wary of abusing the refuse filters too much.

**Comment by ConnorDoyle** [comment]
@adam-mesos we could do that, however I'm concerned about coordinating a reconciliation cycle after each re-registration since handling StatusUpdate messages can become a bottleneck for Marathon.  Thanks for pointing out the potential issue with setting too many filters though, it's good to know.

**Comment by kolloch** [comment]
Hey @adam-mesos, just to save me the time to understand all this: I thought that `declineOffers` without a filters argument would actually use a default filter?

**Comment by adam-mesos** [comment]
@kolloch That's correct. Decline sets a default 5sec filter timeout for offers from that slave.
But why bother setting a filter on each slave as you receive each offer when you could just deactivate your framework's offers entirely until you're ready to get more offers? What you really want is the QuiesceOffers/SUPPRESS call that MESOS-3075/MESOS-3037 propose. You still get all your status updates, can even send framework messages, but you're just not getting any offers until you REVIVE.
This mechanism already exists using DeactivateFramework/ReregisterFramework, and would be easy to implement using pure bindings. I'm not sure if the native libmesos' SchedulerDriver's abort() method does exactly what we want here.

**Comment by kolloch** [comment]
I agree with @ConnorDoyle here, at least in my current state (= tired), I will not able to cover all code paths.

It does sound kind of hacky to me but that is subjective.

**Comment by air** [comment]
Going forward, what should the general advice be to framework authors that want to avoid starvation in the presence of many composing frameworks?

**Comment by kolloch** [comment]
@air, good point!

**Comment by xtfxme** [comment]
@kolloch what was the recommendation? did I miss it somewhere?

We are experiencing marathon offer starvation when running in the precence of multiple instances of our custom framework. Once 5 instances + marathon is running, even if our framework rejects every single offer and never schedules a thing, marathon is unable to ever schedule again, but our frameworks appear unaffected? Still investigating.

I've read some infos:

https://issues.apache.org/jira/browse/MESOS-3202
https://issues.apache.org/jira/browse/MESOS-3037

but any idea the best way to mitigate? Does boosting the filter timeout appear to work? Is it possible to use a massive timeout be default and then `RequestOffer`?

**Comment by xtfxme** [comment]
Specifically, to framework writers and thus marathon, what should be considered when implementing a general purpose framework designed to coexist with other, likely foreign, frameworks? Since marathon is a well-known/reference implementation, I basically just want to copy your current/upcoming strategy.

Right now I think refuse_seconds=3 within our framework, all the time, is killing marathon, but this was really only done to ensure tasks could be started in a timely manner.

If this discussion is better on-list, let me know.
